# Library

In [3]:
import pandas as pd
import numpy as np
import plotly.express as px
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error,mean_squared_error

# Data

Data explored on clustering jupyter file.

In [4]:
data = pd.read_csv('apartments_for_rent_classified_10K.csv', sep = ";", encoding='cp1252')
#C:\Users\ville\Documents\ml_projects\apartment\apartments_for_rent_classified_10K.csv
data

id                category  \
0     5668626895  housing/rent/apartment   
1     5664597177  housing/rent/apartment   
2     5668626833  housing/rent/apartment   
3     5659918074  housing/rent/apartment   
4     5668626759  housing/rent/apartment   
...          ...                     ...   
9995  5630240092  housing/rent/apartment   
9996  5668640983  housing/rent/apartment   
9997  5668643292  housing/rent/apartment   
9998  5668662559  housing/rent/apartment   
9999  5509132540  housing/rent/apartment   

                                                  title  \
0     Studio apartment 2nd St NE, Uhland Terrace NE,...   
1                     Studio apartment 814 Schutte Road   
2     Studio apartment N Scott St, 14th St N, Arling...   
3                        Studio apartment 1717 12th Ave   
4     Studio apartment Washington Blvd, N Cleveland ...   
...                                                 ...   
9995                 Five BR 5407 Abbott Place - Abbott   
9996                            Six BR 256 Las Entradas   
9997                        Six BR 9908 Bentcross Drive   
9998                        One BR in New York NY 10069   
9999         Beautiful Lawrenceville Apartment for rent   

                                                   body  \
0     This unit is located at second St NE, Uhland T...   
1     This unit is located at 814 Schutte Road, Evan...   
2     This unit is located at N Scott St, 14th St N,...   
3     This unit is located at 1717 12th Ave, Seattle...   
4     This unit is located at Washington Blvd, N Cle...   
...                                                 ...   
9995  This unit is located at 5407 Abbott Place - Ab...   
9996  This unit is located at 256 Las Entradas, Mont...   
9997  This unit is located at 9908 Bentcross Drive, ...   
9998  Monthly Rent$4,605 -to $4,790AmenitiesThe Aldy...   
9999  Square footage: 880 sq. feet, unit number: 150...   

                                              amenities  bathrooms  bedrooms  \
0                                                   NaN        NaN       0.0   
1                                                   NaN        NaN       1.0   
2                                                   NaN        1.0       0.0   
3                                                   NaN        1.0       0.0   
4                                                   NaN        NaN       0.0   
...                                                 ...        ...       ...   
9995                                                NaN        4.0       5.0   
9996                                                NaN        8.0       6.0   
9997                                                NaN        8.5       6.0   
9998  Basketball,Cable or Satellite,Doorman,Hot Tub,...        NaN       1.0   
9999                                     Gym,Patio/Deck        1.0       1.0   

     currency fee  has_photo  ... price_display  price_type square_feet  \
0         USD  No  Thumbnail  ...          $790     Monthly         101   
1         USD  No  Thumbnail  ...          $425     Monthly         106   
2         USD  No  Thumbnail  ...        $1,390     Monthly         107   
3         USD  No  Thumbnail  ...          $925     Monthly         116   
4         USD  No  Thumbnail  ...          $880     Monthly         125   
...       ...  ..        ...  ...           ...         ...         ...   
9995      USD  No  Thumbnail  ...        $6,000     Monthly        6300   
9996      USD  No  Thumbnail  ...       $25,000     Monthly        8716   
9997      USD  No  Thumbnail  ...       $11,000     Monthly       11318   
9998      USD  No  Thumbnail  ...        $4,790     Monthly       40000   
9999      USD  No         No  ...        $1,009     Monthly         880   

                        address       cityname state latitude longitude  \
0                           NaN     Washington    DC  38.9057  -76.9861   
1                814 Schutte Rd     Evansville    IN  3

In [5]:
outliers_price = data[data['price']>5500].index
outliers_square_feet = data[data['square_feet']>4000].index
outliers_bedrooms = data[data['bedrooms']>=6].index
outliers_bathrooms = data[data['bathrooms']>3.5].index

outliers = outliers_bedrooms.tolist()+outliers_bathrooms.tolist()+outliers_price.tolist()+outliers_square_feet.tolist()
outliers = set(outliers)
outliers

{5052,
 6257,
 7129,
 7436,
 7806,
 7808,
 7906,
 7907,
 8047,
 8081,
 8206,
 8220,
 8371,
 8477,
 8573,
 8581,
 8707,
 8829,
 8844,
 8887,
 8978,
 9048,
 9128,
 9227,
 9275,
 9276,
 9319,
 9334,
 9347,
 9374,
 9380,
 9402,
 9424,
 9483,
 9485,
 9517,
 9521,
 9527,
 9541,
 9547,
 9567,
 9634,
 9635,
 9639,
 9649,
 9662,
 9685,
 9702,
 9707,
 9725,
 9727,
 9730,
 9731,
 9734,
 9740,
 9749,
 9757,
 9763,
 9770,
 9781,
 9797,
 9810,
 9813,
 9825,
 9840,
 9847,
 9854,
 9861,
 9864,
 9872,
 9879,
 9880,
 9888,
 9892,
 9898,
 9903,
 9904,
 9910,
 9914,
 9917,
 9922,
 9923,
 9924,
 9926,
 9929,
 9932,
 9934,
 9936,
 9938,
 9940,
 9943,
 9945,
 9946,
 9947,
 9954,
 9955,
 9959,
 9964,
 9965,
 9966,
 9967,
 9968,
 9969,
 9970,
 9972,
 9973,
 9974,
 9975,
 9976,
 9977,
 9978,
 9979,
 9980,
 9981,
 9982,
 9983,
 9984,
 9985,
 9986,
 9987,
 9988,
 9989,
 9990,
 9991,
 9992,
 9993,
 9994,
 9995,
 9996,
 9997,
 9998}

In [6]:
data_no_outliers = data.drop(index=outliers, inplace=False)
data_no_outliers.shape

(9869, 22)

## X variables

In [61]:
data.columns[[5,6,13,14,16,17]]

Index(['bathrooms', 'bedrooms', 'price_type', 'square_feet', 'cityname',
       'state'],
      dtype='object')

In [62]:
x_var = data.iloc[:,[5,6,13,14,16,17]].dropna(axis=0).values
x_var

array([[1.0, 0.0, 'Monthly', 107, 'Arlington', 'VA'],
       [1.0, 0.0, 'Monthly', 116, 'Seattle', 'WA'],
       [1.0, 0.0, 'Monthly', 130, 'Manhattan', 'NY'],
       ...,
       [8.0, 6.0, 'Monthly', 8716, 'Montecito', 'CA'],
       [8.5, 6.0, 'Monthly', 11318, 'Potomac', 'MD'],
       [1.0, 1.0, 'Monthly', 880, 'Lawrenceville', 'GA']], dtype=object)

In [63]:
x_var.shape

(9883, 6)

In [64]:
x_var[0]

array([1.0, 0.0, 'Monthly', 107, 'Arlington', 'VA'], dtype=object)

In [65]:
#label_enconder_category     = LabelEncoder()
label_enconder_price_type   = LabelEncoder()
label_enconder_cityname     = LabelEncoder()
label_enconder_state        = LabelEncoder()

#x_var[:,0]  = label_enconder_category.fit_transform(x_var[:,0])
x_var[:,2]  = label_enconder_price_type.fit_transform(x_var[:,2])
x_var[:,4]  = label_enconder_cityname.fit_transform(x_var[:,4])
x_var[:,5]  = label_enconder_state.fit_transform(x_var[:,5])

In [66]:
x_var[0]

array([1.0, 0.0, 0, 107, 52, 45], dtype=object)

### Adding cluster from clustering process

In [68]:
cluster = pd.read_csv('10K_cluster.csv').drop(['Unnamed: 0'], axis=1).rename({'0':'cluster'},axis=1)
cluster

cluster
0           0
1           0
2           0
3           0
4           0
...       ...
9878        1
9879        1
9880        1
9881        1
9882        0

[9883 rows x 1 columns]

In [69]:
x_var

array([[1.0, 0.0, 0, 107, 52, 45],
       [1.0, 0.0, 0, 116, 1283, 47],
       [1.0, 0.0, 0, 130, 820, 34],
       ...,
       [8.0, 6.0, 0, 8716, 920, 4],
       [8.5, 6.0, 0, 11318, 1137, 20],
       [1.0, 1.0, 0, 880, 744, 10]], dtype=object)

In [70]:
x_var = np.c_[x_var,cluster]
x_var

array([[1.0, 0.0, 0, ..., 52, 45, 0],
       [1.0, 0.0, 0, ..., 1283, 47, 0],
       [1.0, 0.0, 0, ..., 820, 34, 0],
       ...,
       [8.0, 6.0, 0, ..., 920, 4, 1],
       [8.5, 6.0, 0, ..., 1137, 20, 1],
       [1.0, 1.0, 0, ..., 744, 10, 0]], dtype=object)

## X variables - no outliers

In [71]:
x_var_out = data_no_outliers.iloc[:,[5,6,13,14,16,17]].dropna(axis=0).values
x_var_out

array([[1.0, 0.0, 'Monthly', 107, 'Arlington', 'VA'],
       [1.0, 0.0, 'Monthly', 116, 'Seattle', 'WA'],
       [1.0, 0.0, 'Monthly', 130, 'Manhattan', 'NY'],
       ...,
       [3.5, 5.0, 'Monthly', 3748, 'Lawrenceville', 'GA'],
       [3.5, 5.0, 'Monthly', 4000, 'Saint Leonard', 'MD'],
       [1.0, 1.0, 'Monthly', 880, 'Lawrenceville', 'GA']], dtype=object)

In [72]:
x_var_out.shape

(9753, 6)

In [73]:
x_var_out[0]

array([1.0, 0.0, 'Monthly', 107, 'Arlington', 'VA'], dtype=object)

In [74]:
#x_var_out[:,0]  = label_enconder_category.fit_transform(x_var_out[:,0])
x_var_out[:,2]  = label_enconder_price_type.fit_transform(x_var_out[:,2])
x_var_out[:,4]  = label_enconder_cityname.fit_transform(x_var_out[:,4])
x_var_out[:,5]  = label_enconder_state.fit_transform(x_var_out[:,2])

In [75]:
x_var_out[0]

array([1.0, 0.0, 0, 107, 52, 0], dtype=object)

In [76]:
x_var_out.shape

(9753, 6)

### Adding cluster from clustering process

In [77]:
cluster = pd.read_csv('10K_cluster_nooutliers.csv').drop(['Unnamed: 0'], axis=1).rename({'0':'cluster'},axis=1)
cluster

cluster
0           0
1           0
2           0
3           0
4           0
...       ...
9748        1
9749        1
9750        1
9751        1
9752        0

[9753 rows x 1 columns]

In [78]:
x_var_out

array([[1.0, 0.0, 0, 107, 52, 0],
       [1.0, 0.0, 0, 116, 1268, 0],
       [1.0, 0.0, 0, 130, 809, 0],
       ...,
       [3.5, 5.0, 0, 3748, 733, 0],
       [3.5, 5.0, 0, 4000, 1216, 0],
       [1.0, 1.0, 0, 880, 733, 0]], dtype=object)

In [79]:
x_var_out = np.c_[x_var_out,cluster]
x_var_out

array([[1.0, 0.0, 0, ..., 52, 0, 0],
       [1.0, 0.0, 0, ..., 1268, 0, 0],
       [1.0, 0.0, 0, ..., 809, 0, 0],
       ...,
       [3.5, 5.0, 0, ..., 733, 0, 1],
       [3.5, 5.0, 0, ..., 1216, 0, 1],
       [1.0, 1.0, 0, ..., 733, 0, 0]], dtype=object)

# Y var - price

In [44]:
data.columns

'price'

In [80]:
y_price = data.dropna(subset=['bathrooms', 'bedrooms', 'price_type', 'square_feet',
                           'cityname', 'state'],axis = 0).iloc[:,11].values #price
y_price

array([ 1390,   925,  2475, ..., 25000, 11000,  1009], dtype=int64)

In [81]:
y_price.shape

(9883,)

# Y var - price - no outlier

In [82]:
x_var_out.shape

(9753, 7)

In [83]:
y_price_out = data_no_outliers.dropna(subset=['bathrooms', 'bedrooms', 'price_type', 'square_feet',
                           'cityname', 'state'],axis = 0).iloc[:,11].values #price
y_price_out

array([1390,  925, 2475, ..., 2295, 2550, 1009], dtype=int64)

In [84]:
y_price_out.shape

(9753,)

# Train and Test Data

1º size: 70%-30% (train-test)

2º size: 80%-20% (train-test)

In [100]:
x_var_train, x_var_test, y_price_train, y_price_test = train_test_split(x_var,y_price,
                                                                        test_size=0.3,
                                                                        random_state=0)
# the best result was with test-size = 30%

In [123]:
x_var_out_train, x_var_out_test, y_price_out_train, y_price_out_test = train_test_split(x_var_out,y_price_out,
                                                                                        test_size=0.2,
                                                                                        random_state=0)
# the best result was with test-size = 20%

# Regression - Decision Tree

In [105]:
from sklearn.tree import DecisionTreeRegressor

## With ouliers

In [101]:
reg_dec_tree = DecisionTreeRegressor()
reg_dec_tree.fit(x_var_train,y_price_train)

DecisionTreeRegressor()

In [102]:
#R^2
reg_dec_tree.score(x_var_train,y_price_train)

0.9966505253973317

In [103]:
#R^2
reg_dec_tree.score(x_var_test,y_price_test)

0.22353691119114416

In [213]:
prediction = reg_dec_tree.predict(x_var_test)
mean_absolute_error(y_price_test,prediction), mean_squared_error(y_price_test,prediction)

(394.4830683369469, 1366430.452783875)

## With NO outliers

In [124]:
reg_dec_tree_out = DecisionTreeRegressor()
reg_dec_tree_out.fit(x_var_out_train,y_price_out_train)

DecisionTreeRegressor()

In [125]:
#R^2
reg_dec_tree_out.score(x_var_out_train,y_price_out_train)

0.9929951350580111

In [126]:
#R^2
reg_dec_tree_out.score(x_var_out_test,y_price_out_test)

0.03687654248211991

In [214]:
prediction = reg_dec_tree_out.predict(x_var_out_test)
mean_absolute_error(y_price_out_test,prediction), mean_squared_error(y_price_out_test,prediction)

(432.2193465043414, 458587.4104642195)

# Regression - Random Forest

In [129]:
from sklearn.ensemble import RandomForestRegressor

## With outliers

In [130]:
reg_random_f= RandomForestRegressor(n_estimators=100)
reg_random_f.fit(x_var_train,y_price_train)

RandomForestRegressor()

In [131]:
#R^2
reg_random_f.score(x_var_train,y_price_train), reg_random_f.score(x_var_test,y_price_test)

(0.9443368984757938, 0.3563644004196447)

In [215]:
prediction = reg_random_f.predict(x_var_test)
mean_absolute_error(y_price_test,prediction), mean_squared_error(y_price_test,prediction)

(325.28563261244085, 1132678.8052624492)

## With NO outliers

In [133]:
reg_random_f_out= RandomForestRegressor(n_estimators=100)
reg_random_f_out.fit(x_var_out_train,y_price_out_train)

RandomForestRegressor()

In [134]:
#R^2
reg_random_f_out.score(x_var_out_train,y_price_out_train), reg_random_f_out.score(x_var_out_test,y_price_out_test)

(0.9223733483073355, 0.4552373319646703)

In [216]:
prediction = reg_random_f_out.predict(x_var_out_test)
mean_absolute_error(y_price_out_test,prediction), mean_squared_error(y_price_out_test,prediction)

(347.70986719138784, 260429.5014446485)

# Regression - SVM

In [136]:
from sklearn.svm import SVR

## Scaling data

In [137]:
from sklearn.preprocessing import StandardScaler

In [ ]:
x_var_train.shape, x_var_test.shape, y_price_train.shape, y_price_test.shape

((6918, 7), (2965, 7), (6918,), (2965,))

In [228]:
scalerX = StandardScaler()
scalerY = StandardScaler()

x_var_scaled_train = scalerX.fit_transform(x_var_train)
x_var_scaled_test = scalerX.transform(x_var_test)
y_price_scaled_train = scalerY.fit_transform(y_price_train.reshape(-1,1))
y_price_scaled_test = scalerY.transform(y_price_test.reshape(-1,1))

scalerX_out = StandardScaler()
scalerY_out = StandardScaler()

x_var_out_scaled_train = scalerX_out.fit_transform(x_var_out_train)
x_var_out_scaled_test = scalerX_out.transform(x_var_out_test)
y_price_out_scaled_train = scalerY_out.fit_transform(y_price_out_train.reshape(-1,1))
y_price_out_scaled_test = scalerY_out.transform(y_price_out_test.reshape(-1,1))

In [229]:
x_var_scaled_train.shape, x_var_scaled_test.shape, y_price_scaled_train.shape, y_price_scaled_test.shape

((6918, 7), (2965, 7), (6918, 1), (2965, 1))

## With outliers

In [230]:
reg_svr_linear = SVR(kernel='linear')
reg_svr_linear.fit(x_var_scaled_train,y_price_scaled_train)

C:\Users\ville\AppData\Roaming\Python\Python311\site-packages\sklearn\utils\validation.py:1300: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVR(kernel='linear')

In [145]:
#R^2
reg_svr_linear.score(x_var_scaled_train,y_price_scaled_train), reg_svr_linear.score(x_var_scaled_test,y_price_scaled_test)

(0.23935034311043302, 0.1281863074260915)

In [252]:
prediction = reg_svr_linear.predict(x_var_scaled_test)
mean_absolute_error(scalerY.inverse_transform(y_price_scaled_test).reshape(1,-1),
                    scalerY.inverse_transform([prediction])), mean_squared_error(scalerY.inverse_transform(y_price_scaled_test).reshape(1,-1),
                                                                                 scalerY.inverse_transform([prediction]))

(490.13561535603156, 1534668.8365777002)

In [253]:
reg_svr_rbf = SVR(kernel='rbf')
reg_svr_rbf.fit(x_var_scaled_train,y_price_scaled_train)

C:\Users\ville\AppData\Roaming\Python\Python311\site-packages\sklearn\utils\validation.py:1300: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVR()

In [254]:
#R^2
reg_svr_rbf.score(x_var_scaled_train,y_price_scaled_train), reg_svr_rbf.score(x_var_scaled_test,y_price_scaled_test)

(0.3491670551664926, 0.19630100737385092)

In [255]:
prediction = reg_svr_rbf.predict(x_var_scaled_test)
mean_absolute_error(scalerY.inverse_transform(y_price_scaled_test).reshape(1,-1),
                    scalerY.inverse_transform([prediction])), mean_squared_error(scalerY.inverse_transform(y_price_scaled_test).reshape(1,-1),
                                                                                 scalerY.inverse_transform([prediction]))

(459.63961602446483, 1414360.571963311)

In [256]:
reg_svr_poly = SVR(kernel='poly')
reg_svr_poly.fit(x_var_scaled_train,y_price_scaled_train)

C:\Users\ville\AppData\Roaming\Python\Python311\site-packages\sklearn\utils\validation.py:1300: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVR(kernel='poly')

In [257]:
#R^2
reg_svr_poly.score(x_var_scaled_train,y_price_scaled_train), reg_svr_poly.score(x_var_scaled_test,y_price_scaled_test)

(0.3794671615193975, 0.16743731188408084)

In [258]:
prediction = reg_svr_poly.predict(x_var_scaled_test)
mean_absolute_error(scalerY.inverse_transform(y_price_scaled_test).reshape(1,-1),
                    scalerY.inverse_transform([prediction])), mean_squared_error(scalerY.inverse_transform(y_price_scaled_test).reshape(1,-1),
                                                                                 scalerY.inverse_transform([prediction]))

(487.27961729212967, 1465155.3013787256)

## With NO outliers

In [259]:
reg_svr_out_linear = SVR(kernel='linear')
reg_svr_out_linear.fit(x_var_out_scaled_train,y_price_out_scaled_train)

C:\Users\ville\AppData\Roaming\Python\Python311\site-packages\sklearn\utils\validation.py:1300: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVR(kernel='linear')

In [260]:
#R^2
reg_svr_out_linear.score(x_var_out_scaled_train,y_price_out_scaled_train), reg_svr_out_linear.score(x_var_out_scaled_test,y_price_out_scaled_test)

(0.148286302231592, 0.165439974174032)

In [261]:
prediction = reg_svr_out_linear.predict(x_var_out_scaled_test)
mean_absolute_error(scalerY.inverse_transform(y_price_out_scaled_test).reshape(1,-1),
                    scalerY.inverse_transform([prediction])), mean_squared_error(scalerY.inverse_transform(y_price_out_scaled_test).reshape(1,-1),
                                                                                 scalerY.inverse_transform([prediction]))

(562.3167730318049, 686816.8012847983)

In [262]:
reg_svr_out_rbf = SVR(kernel='rbf')
reg_svr_out_rbf.fit(x_var_out_scaled_train,y_price_out_scaled_train)

C:\Users\ville\AppData\Roaming\Python\Python311\site-packages\sklearn\utils\validation.py:1300: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVR()

In [263]:
#R^2
reg_svr_out_rbf.score(x_var_out_scaled_train,y_price_out_scaled_train), reg_svr_out_rbf.score(x_var_out_scaled_test,y_price_out_scaled_test)

(0.23804928871444409, 0.22483209475727906)

In [264]:
prediction = reg_svr_out_rbf.predict(x_var_out_scaled_test)
mean_absolute_error(scalerY.inverse_transform(y_price_out_scaled_test).reshape(1,-1),
                    scalerY.inverse_transform([prediction])), mean_squared_error(scalerY.inverse_transform(y_price_out_scaled_test).reshape(1,-1),
                                                                                 scalerY.inverse_transform([prediction]))

(541.9287043223743, 637938.9434696757)

In [265]:
reg_svr_out_poly = SVR(kernel='poly')
reg_svr_out_poly.fit(x_var_out_scaled_train,y_price_out_scaled_train)

C:\Users\ville\AppData\Roaming\Python\Python311\site-packages\sklearn\utils\validation.py:1300: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVR(kernel='poly')

In [266]:
#R^2
reg_svr_out_poly.score(x_var_out_scaled_train,y_price_out_scaled_train), reg_svr_out_poly.score(x_var_out_scaled_test,y_price_out_scaled_test)

(0.18843092172839704, 0.1839260342787863)

In [267]:
prediction = reg_svr_out_poly.predict(x_var_out_scaled_test)
mean_absolute_error(scalerY.inverse_transform(y_price_out_scaled_test).reshape(1,-1),
                    scalerY.inverse_transform([prediction])), mean_squared_error(scalerY.inverse_transform(y_price_out_scaled_test).reshape(1,-1),
                                                                                 scalerY.inverse_transform([prediction]))

(558.9210043138123, 671603.3519502941)

# Regression - Neural Network

In [161]:
from sklearn.neural_network import MLPRegressor

Testing hidden layers= 1, 2 e 10
size 6 e 2470

## With outliers

In [201]:
y_price.shape[0]/4

2470.75

In [268]:
reg_neural = MLPRegressor(max_iter=1000, 
                          random_state=0,
                          tol=0.000001,
                          hidden_layer_sizes=6) #one layer
reg_neural.fit(x_var_scaled_train,y_price_scaled_train)

C:\Users\ville\AppData\Roaming\Python\Python311\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1624: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


MLPRegressor(hidden_layer_sizes=6, max_iter=1000, random_state=0, tol=1e-06)

In [269]:
#R^2
reg_neural.score(x_var_scaled_train,y_price_scaled_train), reg_neural.score(x_var_scaled_test,y_price_scaled_test)

(0.4250632314536661, 0.23369223356668634)

In [271]:
prediction = reg_neural.predict(x_var_scaled_test)
mean_absolute_error(scalerY.inverse_transform(y_price_scaled_test).reshape(1,-1),
                    scalerY.inverse_transform([prediction])), mean_squared_error(scalerY.inverse_transform(y_price_scaled_test).reshape(1,-1),
                                                                                 scalerY.inverse_transform([prediction]))

(477.9865210230854, 1348558.9764036315)

In [272]:
reg_neural = MLPRegressor(max_iter=1000, 
                          random_state=0,
                          tol=0.000001,
                          hidden_layer_sizes=np.repeat(6,10))
reg_neural.fit(x_var_scaled_train,y_price_scaled_train)

C:\Users\ville\AppData\Roaming\Python\Python311\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1624: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


MLPRegressor(hidden_layer_sizes=array([6, 6, 6, 6, 6, 6, 6, 6, 6, 6]),
             max_iter=1000, random_state=0, tol=1e-06)

In [273]:
#R^2
reg_neural.score(x_var_scaled_train,y_price_scaled_train), reg_neural.score(x_var_scaled_test,y_price_scaled_test)

(0.3994871975670262, 0.20238053420860525)

In [274]:
prediction = reg_neural.predict(x_var_scaled_test)
mean_absolute_error(scalerY.inverse_transform(y_price_scaled_test).reshape(1,-1),
                    scalerY.inverse_transform([prediction])), mean_squared_error(scalerY.inverse_transform(y_price_scaled_test).reshape(1,-1),
                                                                                 scalerY.inverse_transform([prediction]))

(478.7022312712349, 1403661.7367897439)

In [275]:
reg_neural = MLPRegressor(max_iter=1000, 
                          random_state=0,
                          tol=0.000001,
                          hidden_layer_sizes=2470)
reg_neural.fit(x_var_scaled_train,y_price_scaled_train)

C:\Users\ville\AppData\Roaming\Python\Python311\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1624: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


MLPRegressor(hidden_layer_sizes=2470, max_iter=1000, random_state=0, tol=1e-06)

In [276]:
#R^2
reg_neural.score(x_var_scaled_train,y_price_scaled_train), reg_neural.score(x_var_scaled_test,y_price_scaled_test)

(0.5810826066358149, 0.3037496527088499)

In [277]:
prediction = reg_neural.predict(x_var_scaled_test)
mean_absolute_error(scalerY.inverse_transform(y_price_scaled_test).reshape(1,-1),
                    scalerY.inverse_transform([prediction])), mean_squared_error(scalerY.inverse_transform(y_price_scaled_test).reshape(1,-1),
                                                                                 scalerY.inverse_transform([prediction]))

(438.4099481676616, 1225270.9639545784)

In [278]:
reg_neural = MLPRegressor(max_iter=1000, 
                          random_state=0,
                          tol=0.000001,
                          hidden_layer_sizes=(2470,6))
reg_neural.fit(x_var_scaled_train,y_price_scaled_train)

C:\Users\ville\AppData\Roaming\Python\Python311\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1624: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


MLPRegressor(hidden_layer_sizes=(2470, 6), max_iter=1000, random_state=0,
             tol=1e-06)

In [279]:
#R^2
reg_neural.score(x_var_scaled_train,y_price_scaled_train), reg_neural.score(x_var_scaled_test,y_price_scaled_test)

(0.5670400291158098, 0.2696606131703492)

In [280]:
prediction = reg_neural.predict(x_var_scaled_test)
mean_absolute_error(scalerY.inverse_transform(y_price_scaled_test).reshape(1,-1),
                    scalerY.inverse_transform([prediction])), mean_squared_error(scalerY.inverse_transform(y_price_scaled_test).reshape(1,-1),
                                                                                 scalerY.inverse_transform([prediction]))

(453.6119308879051, 1285261.3258956955)

## With NO outliers

In [202]:
y_price_out.shape[0]/4

2438.25

In [281]:
reg_neural_out = MLPRegressor(max_iter=1000,
                              random_state=0,
                              tol=0.000001,
                              hidden_layer_sizes=6) #one layer
reg_neural_out.fit(x_var_out_scaled_train,y_price_out_scaled_train)

C:\Users\ville\AppData\Roaming\Python\Python311\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1624: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


MLPRegressor(hidden_layer_sizes=6, max_iter=1000, random_state=0, tol=1e-06)

In [282]:
#R^2
reg_neural_out.score(x_var_out_scaled_train,y_price_out_scaled_train), reg_neural_out.score(x_var_out_scaled_test,y_price_out_scaled_test)

(0.24759388176188113, 0.23661432419374273)

In [283]:
prediction = reg_neural_out.predict(x_var_out_scaled_test)
mean_absolute_error(scalerY.inverse_transform(y_price_out_scaled_test).reshape(1,-1),
                    scalerY.inverse_transform([prediction])), mean_squared_error(scalerY.inverse_transform(y_price_out_scaled_test).reshape(1,-1),
                                                                                 scalerY.inverse_transform([prediction]))

(583.0909212184181, 628242.5371200586)

In [284]:
reg_neural_out = MLPRegressor(max_iter=1000,
                              random_state=0,
                              tol=0.000001,
                              hidden_layer_sizes=np.repeat(6,10)) #10 layer
reg_neural_out.fit(x_var_out_scaled_train,y_price_out_scaled_train)

C:\Users\ville\AppData\Roaming\Python\Python311\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1624: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


MLPRegressor(hidden_layer_sizes=array([6, 6, 6, 6, 6, 6, 6, 6, 6, 6]),
             max_iter=1000, random_state=0, tol=1e-06)

In [285]:
#R^2
reg_neural_out.score(x_var_out_scaled_train,y_price_out_scaled_train), reg_neural_out.score(x_var_out_scaled_test,y_price_out_scaled_test)

(0.2501591344257288, 0.23959089727056637)

In [286]:
prediction = reg_neural_out.predict(x_var_out_scaled_test)
mean_absolute_error(scalerY.inverse_transform(y_price_out_scaled_test).reshape(1,-1),
                    scalerY.inverse_transform([prediction])), mean_squared_error(scalerY.inverse_transform(y_price_out_scaled_test).reshape(1,-1),
                                                                                 scalerY.inverse_transform([prediction]))

(584.8991496123647, 625792.9105669641)

In [287]:
reg_neural_out = MLPRegressor(max_iter=1000,
                              random_state=0,
                              tol=0.000001,
                              hidden_layer_sizes=2438) #1 layer
reg_neural_out.fit(x_var_out_scaled_train,y_price_out_scaled_train)

C:\Users\ville\AppData\Roaming\Python\Python311\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1624: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


MLPRegressor(hidden_layer_sizes=2438, max_iter=1000, random_state=0, tol=1e-06)

In [288]:
#R^2
reg_neural_out.score(x_var_out_scaled_train,y_price_out_scaled_train), reg_neural_out.score(x_var_out_scaled_test,y_price_out_scaled_test)

(0.2776210593054437, 0.26790164589741716)

In [289]:
prediction = reg_neural_out.predict(x_var_out_scaled_test)
mean_absolute_error(scalerY.inverse_transform(y_price_out_scaled_test).reshape(1,-1),
                    scalerY.inverse_transform([prediction])), mean_squared_error(scalerY.inverse_transform(y_price_out_scaled_test).reshape(1,-1),
                                                                                 scalerY.inverse_transform([prediction]))

(565.9820086121796, 602494.0498353735)

In [290]:
reg_neural_out = MLPRegressor(max_iter=1000,
                              random_state=0,
                              tol=0.000001,
                              hidden_layer_sizes=(2438,6)) #2 layers
reg_neural_out.fit(x_var_out_scaled_train,y_price_out_scaled_train)

C:\Users\ville\AppData\Roaming\Python\Python311\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1624: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


MLPRegressor(hidden_layer_sizes=(2438, 6), max_iter=1000, random_state=0,
             tol=1e-06)

In [291]:
#R^2
reg_neural_out.score(x_var_out_scaled_train,y_price_out_scaled_train), reg_neural_out.score(x_var_out_scaled_test,y_price_out_scaled_test)

(0.28901686487820244, 0.2614104646097791)

In [292]:
prediction = reg_neural_out.predict(x_var_out_scaled_test)
mean_absolute_error(scalerY.inverse_transform(y_price_out_scaled_test).reshape(1,-1),
                    scalerY.inverse_transform([prediction])), mean_squared_error(scalerY.inverse_transform(y_price_out_scaled_test).reshape(1,-1),
                                                                                 scalerY.inverse_transform([prediction]))

(573.9626437594684, 607836.0890303649)

# Regression - Final model

## $R^2$

*" It measures how well the model fits the data by assessing the proportion of variance in the dependent variable explained by the independent variables"*

Results for data with outliers:

$R^2$:

- Decision Tree: 0.22353691119114416

- Random Forest: 0.3563644004196447

- SVR:

    - Linear: 0.1281863074260915

    - RBF: 0.19630100737385092

    - Poly: 0.16743731188408084

- Neural Network: best $R^2$: 0.3037496527088499

Final best model, based on $R^2$, is **Random Forest**

Results for data with **NO** outliers:

$R^2$:

- Decision Tree: 0.03687654248211991

- Random Forest: 0.4552373319646703

- SVR:

    - Linear: 0.165439974174032

    - RBF: 0.22483209475727906

    - Poly: 0.1839260342787863

- Neural Network: best $R^2$: 0.26790164589741716

Final best model, based on $R^2$, is **Random Forest**

## MAE - Mean Absolute Error

*"Simple metric which calculates the absolute difference between actual and predicted values."*


Results for data with outliers:

**MAE**:

- Decision Tree: 394.4830683369469

- Random Forest: 325.28563261244085

- SVR:

    - Linear: 490.13561535603156

    - RBF: 459.63961602446483

    - Poly: 487.27961729212967

- Neural Network: 438.4099481676616 (from the best $R^2$ model)

Final best model is **Random Forest**

Results for data with **NO** outliers:

**MAE**:

- Decision Tree: 432.2193465043414

- Random Forest: 347.70986719138784

- SVR:

    - Linear: 562.3167730318049

    - RBF: 541.9287043223743

    - Poly: 558.9210043138123

- Neural Network: 565.9820086121796 (from the best $R^2$ model)

Final best model is **Random Forest**

## MSE - Mean Squared Error

*"It represents the squared distance between actual and predicted values, we perform squared to avoid the cancellation of negative terms and it is the benefit of MSE"*

Results for data with outliers:

**MSE**:

- Decision Tree: 1,366,430.452783875

- Random Forest: 1,132,678.8052624492

- SVR:

    - Linear: 1,534,668.8365777002

    - RBF: 1,414,360.571963311

    - Poly: 1,465,155.3013787256

- Neural Network: 1,225,270.9639545784 (from the best $R^2$ model)

Final best model is **Random Forest**

Results for data with **NO** outliers:

**MSE**:

- Decision Tree: 458,587.4104642195

- Random Forest: 260,429.5014446485

- SVR:

    - Linear: 686,816.8012847983

    - RBF: 637,938.9434696757

    - Poly: 671,603.3519502941

- Neural Network: 602,494.0498353735 (from the best $R^2$ model)

Final best model is **Random Forest**

## Conclusion - Result

Between Random Forest models with full data and without outliers, the results **without outliers had the best performance** (based on $R^2$, MAE and MSE).

Note: That MSE with outliers vs. MSE with no outliers are very differente, this is due to outliers - as expected. So MSE is good to compare models intra with ou without outliers but no good to compare models with vs. models without

In [293]:
reg_random_f_out= RandomForestRegressor(n_estimators=100, random_state=0)
reg_random_f_out.fit(x_var_out_train,y_price_out_train)
reg_random_f_out.score(x_var_out_test,y_price_out_test)

0.45304699573605534

# Test with cluster variable vs. without it

In [295]:
x_var_out_nocluster = data_no_outliers.iloc[:,[5,6,13,14,16,17]].dropna(axis=0).values
x_var_out_nocluster[:,2]  = label_enconder_price_type.fit_transform(x_var_out_nocluster[:,2])
x_var_out_nocluster[:,4]  = label_enconder_cityname.fit_transform(x_var_out_nocluster[:,4])
x_var_out_nocluster[:,5]  = label_enconder_state.fit_transform(x_var_out_nocluster[:,2])

x_var_out_nocluster.shape

(9753, 6)

In [296]:
x_var_out_nocluster_train, x_var_out_nocluster_test, y_price_out_train, y_price_out_test = train_test_split(x_var_out_nocluster,
                                                                                        y_price_out,
                                                                                        test_size=0.2,
                                                                                        random_state=0)
# the best result was with test-size = 20%

In [299]:
reg_random_f_out_nocluster= RandomForestRegressor(n_estimators=100, random_state=0)
reg_random_f_out_nocluster.fit(x_var_out_nocluster_train,y_price_out_train)
reg_random_f_out_nocluster.score(x_var_out_nocluster_test,y_price_out_test)

0.43800866404637473

Cluster variable **improved** regression results for price prediction.